src/gdb_packet.c

```diff
@@ -209,165 +245,210 @@ size_t gdb_getpacket(char *const packet, const size_t size)
+			data_size = MIN(data_size, (GDB_PACKET_BUFFER_SIZE - preamble_size) / 2U);
+			hexify(packet_buffer.data + packet_buffer.size, data, data_size);
+			packet_buffer.size += data_size * 2U;
+		} else {
+			data_size = MIN(data_size, GDB_PACKET_BUFFER_SIZE - preamble_size);
+			memcpy(packet_buffer.data + preamble_size, data, data_size);
+			packet_buffer.size += data_size;
+		}
+	}
+
+	/* Transmit the packet */
+	gdb_packet_send(&packet_buffer);
 }
 
 void gdb_putpacket_f(const char *const fmt, ...)
 {
-	va_list ap;
-	char *buf = NULL;
+	/*
+	 * Create a packet using the internal packet buffer
+	 * This destroys the previous packet in the buffer
+	 * any packets obtained from gdb_packet_receive() will be invalidated
+	 */
+	packet_buffer.notification = false;
 
-	va_start(ap, fmt);
-	const int size = vasprintf(&buf, fmt, ap);
-	if (size < 0) {
-		/* Heap exhaustion. Report with puts() elsewhere. */
-		DEBUG_ERROR("gdb_putpacket_f: vasprintf failed\n");
-	} else {
-		gdb_putpacket(buf, size);
-		free(buf);
-	}
+	/*
+	 * Format the string directly into the packet buffer
+	 * This considers GDB_PACKET_BUFFER_SIZE to be the maximum size of the string
+	 * But it does not take into consideration the extra space needed for escaping
+	 * This is safe because the escaping is done during the actual packet transmission
+	 * but it will result in a packet larger than what we told GDB we could handle
+	 */
+	va_list ap;
+	packet_buffer.size = vsnprintf(packet_buffer.data, GDB_PACKET_BUFFER_SIZE + 1U, fmt, ap);
 	va_end(ap);
+
+	/* Transmit the packet */
+	gdb_packet_send(&packet_buffer);
+}
+
+void gdb_put_notificationz(const char *const str)
+{
+	/*
+	 * Create a packet using the internal packet buffer
+	 * This destroys the previous packet in the buffer
+	 * any packets obtained from gdb_packet_receive() will be invalidated
+	 */
+	packet_buffer.notification = true;
+
+	packet_buffer.size = strnlen(str, GDB_PACKET_BUFFER_SIZE);
+	memcpy(packet_buffer.data, str, packet_buffer.size);
+
+	/* Transmit the packet */
+	gdb_packet_send(&packet_buffer);
 }
 
-void gdb_out(const char *const buf)
+void gdb_out(const char *const str)
 {
-	const size_t buf_len = strlen(buf);
-	char *hexdata = calloc(1, 2U * buf_len + 1U);
-	if (!hexdata)
-		return;
-
-	hexify(hexdata, buf, buf_len);
-	gdb_putpacket2("O", 1, hexdata, 2U * buf_len);
-	free(hexdata);
+	/**
+     * Program console output packet
+     * See https://sourceware.org/gdb/current/onlinedocs/gdb.html/Stop-Reply-Packets.html#Stop-Reply-Packets
+     * 
+     * Format; ‘O XX…’
+     * ‘XX…’ is hex encoding of ASCII data, to be written as the program’s console output.
+     * 
+     * Can happen at any time while the program is running and the debugger should continue to wait for ‘W’, ‘T’, etc.
+     * This reply is not permitted in non-stop mode.
+     */
+	gdb_putpacket("O", 1U, str, strnlen(str, GDB_OUT_PACKET_MAX_SIZE), true);
```

Comment:
Might be worth putting this in a loop to chew through GDB_OUT_PACKET_MAX_SIZE bytes at a time, so we can never wind up with a truncated string sent, something like this:

```suggestion
	const size_t str_len = strlen(str);
	for (size_t offset = 0; offset < str_len; offset += GDB_OUT_PACKET_MAX_SIZE) {
		const size_t amount = MIN(str_len - offset, GDB_OUT_PACKET_MAX_SIZE);
		gdb_putpacket("O", 1U, str + offset, amount, true);
	}
```

Reply:
That is a good idea that I too had cross my mind.
For `gdb_out` it's easy enough to do as the string is static and already stored somewhere, but for `gdb_voutf`/`gdb_outf` it's not so simple, we need to allocate space for the formatted string somewhere, and right now its thankfully bounded to half the size of `GDB_OUT_PACKET_MAX_SIZE` but if we allow arbitrarily sized strings there we would need dynamic memory again.
My idea for removing the need for the scratch buffer wouldn't work either as I don't think it's easy to partially format strings (If we hooked into the "write" call of the print f we could do it, but sounds like more trouble than it's worth).

Reply:
I wouldn't do it in one but not the others as that would probably be confusing, having functions that appear similar on the surface have subtly different behaviors

Reply:
🤔 The target.xml flows via the `gdb_putpacket*()` functions, so it doesn't hurt that.. this only impacts strings being sent to GDB's console.. oki!

Please make a note of this limitation then in the header as it applies to all GDB console output functions, and note the default limit in the process please so it's explicit, and we'll mark this resolved.